### ISSUE TITLE
Meteor 1.6.0.1 upgrade to 1.6.1 downgrades to broken packages

### ISSUE THREAD

**Issue by xet7** [meteor/meteor#9609]
Specifying package version does not seem to be possible, causing dependency hell.
I have tried many workarounds without success, me and many others have same problem.
More details at https://github.com/peerlibrary/meteor-computed-field/issues/13

**Comment by spencern** [comment]
I've seen this in our upgrade to 1.6.1 for Reaction Commerce as well.

```
...
check                      upgraded from 1.2.5 to 1.3.0
coffeescript               downgraded from 1.12.7_3 to 1.0.17
coffeescript-compiler      removed from your project
ddp-client                 upgraded from 2.2.0 to 2.3.1
...
```

These are the only two packages that either downgraded or were removed.
We're not seeing any errors because of this at the moment.

More details on our upgrade PR: https://github.com/reactioncommerce/reaction/pull/3615

**Comment by maxnowack** [comment]
I've got the same issue. This prevents me from upgrading to v1.6.1 …

**Comment by wemiced** [comment]
Hi, same issue for me

**Comment by WayneUong** [comment]
My workaround is getting all indirect dependencies to the lowest versions possible and hoping they don't crash.

**Comment by krishaamer** [comment]
Same problem here.

**Comment by ruerob** [comment]
I've got that problem, too!

**Comment by gitTerebi** [comment]
Problem causing coffeescript to downgrade itself, which in turns forces some other packages to downgrade.

**Comment by xet7** [comment]
For peerlibrary I added pull request that fixes this by downgrading coffeescript to 1.0.17 etc. https://github.com/peerlibrary/meteor-computed-field/pull/14

It makes possible for Wekan to upgrade to Meteor 1.6.1.

**Comment by xet7** [comment]
That pull request does not solve anything. Real fix would be to have all dependencies use CoffeeScript 2.

**Comment by xet7** [comment]
When upgrading to Meteor 1.6.1 there also seems to be 100% CPU usage, maybe caused by downgrade of coffeescript or some other change.

**Comment by hwillson** [comment]
**TL;DR**: This issue is caused by outdated 3rd party packages. Those packages need to be updated to work with Meteor 1.6.1 (upstream or through local copies), or alternative packages/solutions need to be found. 

---

The [Meteor 1.6.1 release blog post](https://blog.meteor.com/announcing-meteor-1-6-1-50aad71da4e6) explains why this is happening in detail. I'll quote the relevant section here for those that haven't seen it:

> The major version of the Meteor babel-compiler package has been bumped to 7.0.0, which may be incompatible with existing Meteor packages that explicitly depend on an earlier version of babel-compiler. Version constraints in Meteor typically specify a minimum version, and otherwise constrain only the major version, allowing the minor and patch versions to change. In this case, however, we’ve changed the major version of babel-compiler, which is likely to produce some conflicts.
>
> If you are the author of a Meteor package that depends on babel-compiler, and you have determined that the update is compatible with your package’s previous usage of babel-compiler (give or take any tweaks you might need to make), you can make your package compatible with either Babel 6 or Babel 7 by using the || version constraint syntax:
>
> ```
> Package.onUse(function (api) {
>   ...
>   api.use("babel-compiler@6.19.4||7.0.0");
>   ...
> });
> ```
>
> This approach has been working well for the coffeescript-compiler package, for example. Speaking of CoffeeScript, while Meteor 1.6.1 was under development, a new major version of the coffeescript and coffeescript-compiler packages were released (most recently 2.0.3_4), which is worth mentioning for the same reason as the babel-compiler@7.0.0 upgrade: packages like practicalmeteor:mocha that depended on a 1.x version of coffeescript may need to be updated or replaced. Specifically, we have found meteortesting:mocha to be a good choice.
>
> Unfortunately, conflicting version constraints may originate from Meteor packages that are not actively maintained, in which case you may need to take matters into your own hands. Fortunately Meteor makes this possible: simply clone the package into your local packages/ directory and then modify the code as necessary, likely by relaxing version constraints in its package.js file, but also by making other necessary changes to the package. Your local copy of the package will take precedence over the Atmosphere version.
>
> If your changes seem like they should become part of the original package, please submit a pull request to the upstream project. If the package author is not responsive, it may be time to find an alternative package that is more actively supported. However, this technique of modifying a local clone of a package is a powerful way to solve a variety of problems with Meteor packages that you didn’t write.

In the name of forward progress, we'll close this issue for now. Thanks!

**Comment by xet7** [comment]
@hwillson 

Closing this issue solves nothing. When I define different version of dependency, meteor command does not let it happen in any way, it changes to those versions it wants. There is no possibility to define dependency versions myself.

**Comment by xet7** [comment]
@hwillson

Or maybe I did not read your issue comment well enough. Is there a way to force dependency? For example CoffeeScript 2 ?

**Comment by xet7** [comment]
Nevermind, I think I have enough info about upgrading already, I'll just gather some courage and then try to push trough the brick wall again.

**Comment by hwillson** [comment]
Ha - yes I know it's a pain, but our hands are tied here a bit. We want to move core ahead, but that unfortunately means causing a bit of pain for people that are using out of date or long abandoned packages.

In answer to your question about forcing dependencies, there isn't currently a way to do this exactly. That feature is coming at some point in the future though, and is being tracked here: https://github.com/meteor/meteor-feature-requests/issues/208

There are ways to force certain version ranges though, by using `@~` based version constraints for example (see https://github.com/meteor/meteor/pull/8991 for more info).

The above being said, forcing `coffeescript` 2.x could prove to be difficult, if your application is using packages that are dependent on `coffeescript` 1. Trying to force `coffeescript` 2 in that case will cause version constraint conflicts. Looking at `practicalmeteor:mocha` for example, here's a small [snippet from its `package.js`](https://github.com/practicalmeteor/meteor-mocha/blob/3a2658070a920f8846df48bb8d8c7b678b8c6870/package.js#L27-L31):

```js
Package.onUse(function (api) {
  api.versionsFrom("1.3");
  ...
  api.use('coffeescript');
```

It's using `coffeescript` without an explicit version constraint beside it, but it's making an `api.versionsFrom("1.3");` call. This means there is an exact package constraint on the version of `coffeescript` being used, to be the version that shipped with Meteor 1.3. So when using `practicalmeteor:mocha`, trying to force a `coffeescript` 2 constraint will never work (without package modifications).

Just for completeness sake, here's a quick summary of how Meteor's core package version constraints have worked over the years (extracted from https://github.com/meteor/meteor/pull/8960#issuecomment-322511317):

> - Before Meteor 1.4, core package constraints were exact (package@=a.b.c), which was way too restrictive.
> - From Meteor 1.4-1.4.2.7 there were no constraints, which was better than exact, but bad in other ways. For example, Meteor could never force a core package to upgrade with a new Meteor release. We're still seeing problems due to this policy, as we release new versions of core packages and people still using 1.4.2.x accidentally upgrade to the latest version.
> - From Meteor 1.4.3 until 1.5.1 (current), you could use any >= version with the same major version. This is how package@a.b.c version constraints work. To be honest, I was surprised to realize (recently) that the minor versions were not constrained, which is why Meteor 1.5.2 will use the package@~a.b.c syntax (#8991).
> - Starting in Meteor 1.5.2 and 1.6, the rules are just like 1.5.1, except that the minor versions of core packages also have to match. That's what the new package@~a.b.c syntax means.

So long story short, we're closing this issue as the reason for package downgrades is known (it's being caused by Meteor 1.6.1's major version bump on `babel-compiler` / `coffeescript` and outdated 3rd party packages). Fixing those 3rd party packages is up to individual developers or the Meteor community as a whole. Thanks!

**Comment by xet7** [comment]
@hwillson 

Yes I will of course absolutely upgrade. I'm so exited too about reducing bundle sizes dramatically, because some do import huge Trello boards with attachments to Wekan, so reducing bundle size speeds up Wekan and makes it possible to handle even larger amounts of data.

Thanks a lot for all the improvements to Meteor to all Meteor contributors !

**Comment by maxfi** [comment]
@hwillson Thanks for the detailed info! Just to clarify, are you saying that there is no way to update to Meteor 1.6.1 if there is a dependency on `coffeescript-compiler 1.x` either by the app itself or any dependencies? I'm getting the following error when trying to update:

```
=> Errors while initializing project:         

While selecting package versions:
error: Conflict: Constraint babel-compiler@6.19.4 is not satisfied by babel-compiler 7.0.6.
Constraints on package "babel-compiler":
* babel-compiler@~7.0.0 <- top level
* babel-compiler@7.0.0 <- ecmascript 0.10.0
* babel-compiler@7.0.0 <- minifier-js 2.3.1 <- standard-minifier-js 2.3.1
* babel-compiler@6.19.4 <- coffeescript-compiler 1.12.7_3 <- coffeescript 1.12.7_3
* babel-compiler@7.0.0 <- standard-minifier-js 2.3.1
```

I'm hoping there's a workaround as most of my app is written in Coffeescript v1 and I was planning on refactoring to ECMAScript as I go, rather than refactoring to Coffeescript v2 just to update to Meteor 1.6.1, and then having to refactor to ECMAScript later. Fingers crossed! Thanks.
